### PR TITLE
Audio voice files

### PR DESCRIPTION
**Pull request by cammellos**
https://github.com/status-im/specs/pull/138 for the specs

## Why make this change?
    
We are adding support of audio recorded files, similarly to how we did
    with images
    
## What has changed?
    
- Added protobuf definition, only AAC supported
- Added migrations to store files
- Fixed an issue with nil pointer when transaction would fail to be created, causing the application to crash

### CODE REVIEW

**Comment by cammellos** [comment]
@0kok0 adding you as I can't find a tag for security

File: protocol/protobuf/chat_message.pb.go [hunk 1]
```diff
@@ -350,10 +426,16 @@ type ChatMessage_Image struct {
...
 
+type ChatMessage_Audio struct {
+	Audio *AudioMessage `protobuf:"bytes,11,opt,name=audio,proto3,oneof"`
+}
+
 func (*ChatMessage_Sticker) isChatMessage_Payload() {}
 
 func (*ChatMessage_Image) isChatMessage_Payload() {}
 
+func (*ChatMessage_Audio) isChatMessage_Payload() {}
```
Comment (jakubgs) [inline protocol/protobuf/chat_message.pb.go:448]: What is this and how does this work?

**Review by jakubgs — APPROVED** [review]
I don't see anything wrong with this.

File: protocol/protobuf/chat_message.pb.go (hunk 1 above)
Comment (cammellos) [reply protocol/protobuf/chat_message.pb.go:448]: That's magic, it's true meaning has been lost in the mists of time.
Some say that reciting that magic incantation on a full moon will create a path between you and another being, and through that path they'll hear your voice (after clicking on an icon).
Others say that this is just what a compiled protobuf declaration looks like, and that is automatically generated, so not to be questioned.
It is up to you who to believe.

Comment (jakubgs) [reply protocol/protobuf/chat_message.pb.go:448]: Weird but okay...

**Comment by 0kok0** [comment]
Still reviewing this - don't merge pls

**Review by 0kok0 — APPROVED** [review]
Looks good to me

File: protocol/message.go [hunk 2]
```diff
@@ -253,6 +261,37 @@ func (m *Message) parseImage() error {
 	return nil
 }
 
+// parseAudio check the message contains an audio, and if so
+// it creates the a base64 encoded version of it.
```
Comment (Samyoul) [inline protocol/message.go]: tiny tpyo

```suggestion
// it creates a base64 encoded version of it.
```

File: protocol/persistence.go [hunk 3]
```diff
@@ -31,6 +31,10 @@ func (db sqlitePersistence) SaveChat(chat Chat) error {
 
 func (db sqlitePersistence) SaveChats(chats []*Chat) error {
 	tx, err := db.db.BeginTx(context.Background(), &sql.TxOptions{})
+	if err != nil {
+		return err
+	}
+
```
Comment (Samyoul) [inline protocol/persistence.go:38]: Nice early return

File: protocol/messenger.go [hunk 4]
```diff
@@ -1412,6 +1413,26 @@ func (m *Messenger) SendChatMessage(ctx context.Context, message *Message) (*Mes
 		message.Payload = &protobuf.ChatMessage_Image{Image: &image}
 
 	}
+
+	if len(message.AudioPath) != 0 {
+		file, err := os.Open(message.AudioPath)
```
Comment (Samyoul) [inline protocol/messenger.go:1388]: Is it possible for `if len(message.ImagePath) != 0` and `if len(message.AudioPath) != 0` to both be true?

File: protocol/message_validator.go [hunk 5]
```diff
@@ -209,5 +209,22 @@ func ValidateReceivedChatMessage(message *protobuf.ChatMessage, whisperTimestamp
...
+			return errors.New("no audio content")
+		}
+		if len(audio.Payload) == 0 {
+			return errors.New("audio payload empty")
+		}
+
+		if audio.Type == protobuf.AudioMessage_UNKNOWN_AUDIO_TYPE {
+			return errors.New("audio type unknown")
+		}
+	}
```
Comment (Samyoul) [inline protocol/message_validator.go:227]: This looks like it should be a switch statement on `message.ContentType` with dedicated type validator functions or errors for

- `protobuf.ChatMessage_UNKNOWN_CONTENT_TYPE`
- `protobuf.ChatMessage_TRANSACTION_COMMAND`
- `protobuf.ChatMessage_STICKER`
- `protobuf.ChatMessage_IMAGE`
- `protobuf.ChatMessage_AUDIO`

File: protocol/messenger.go (hunk 4 above)
Comment (Samyoul) [reply protocol/messenger.go:1388]: Looks like these tests are also sub-functions, would it make sense to break them out into their own dedicated function?

File: protocol/message_validator.go (hunk 5 above)
Comment (jakubgs) [reply protocol/message_validator.go:227]: if you're going that far then you might as well make the handler of each type it's own function to keep the `switch` concise.

**Pull request merged**